Initial commit of kafka server configuration

jvlcek opened the pull request:
This PR will automate the steps required to configre kafka messaging service

Fixes https://github.com/ManageIQ/manageiq-appliance_console/issues/125

jvlcek commented:
@miq-bot add_label wip
@miq-bot add_label enhancement

Fryguy commented on lib/manageiq/appliance_console/message_server_configuration.rb:
```diff
@@ -0,0 +1,247 @@
+      :messaging_yaml_sample_path, :messaging_yaml_path
+
+    def initialize(options = {})
+      puts("JJV 0.0 #{File.basename(__FILE__)} / #{__method__}")
```
Just pinging this to be removed later.

Fryguy commented on lib/manageiq/appliance_console/message_server_configuration.rb:
```diff
@@ -0,0 +1,247 @@
+    def activate
+      puts("JJV 0.0 #{File.basename(__FILE__)} / #{__method__}")
+      begin
+        configure_log3               ## STEP 3: Create the log4 configuration file
```
minor but the method is named log3, but the feature is log4?

jvlcek replied:
renamed to create_tools_logconfig

Fryguy commented on lib/manageiq/appliance_console/message_server_configuration.rb:
```diff
@@ -0,0 +1,247 @@
+
+      AwesomeSpawn.run!("firewall-cmd --add-port=9092/tcp --permanent") # insecure
+      AwesomeSpawn.run!("firewall-cmd --add-port=9093/tcp --permanent") # secure
+      AwesomeSpawn.run!("firewall-cmd --reload")
```
Do we do the same thing for postgres?  that is, configure the firewall on the fly?  if so, then :+1:

jvlcek replied:
@Fryguy It doesn't appear that we do firewwall config on the fly for postgres. @bdunne had suggested this approach for kafak. @bdunne Can you please comment?

Fryguy commented on lib/manageiq/appliance_console/message_server_configuration.rb:
```diff
@@ -0,0 +1,247 @@
+      content = <<~EOS
+
+        ssl.keystore.location="#{keystore_path}"
+        ssl.keystore.password="#{@password}"
```
Do we have to handle escaping or anything in any of these passwords (here and elsewhere we do interpolation)?  Like, what if someone has a `"` in their password?

jvlcek replied:
@Fryguy I'll investigate. Thank you!

Fryguy commented:
Looking good @jvlcek !

jvlcek commented:
@miq-bot remove_label wip

@Fryguy and @bdunne Please review.

jvlcek commented:
Regarding codeclimate and hakiri failures: @Fryguy has told me:

"you can ignore codeclimate...it's been having issues on that repo and I can't turn it off
same with hakiri"

bdunne commented on bin/appliance_console:
```diff
@@ -320,6 +320,33 @@ Static Network Configuration
         end
         press_any_key
 
+      when I18n.t("advanced_settings.message_server")
```
I think these should live under https://github.com/ManageIQ/manageiq-appliance_console/pull/130/files#diff-e8b26cb149425ac9294176fd67aa5744bcd030ff9ad96fe836bbdfa50607ce0dR449 with the existing external messaging configuration.

bdunne replied:
Better link... https://github.com/ManageIQ/manageiq-appliance_console/blob/481d9b37c73a44471b13ce5c5d7a8fce00b7429d/bin/appliance_console#L449

jvlcek replied:
@bdunne Thank you. I will rearrange the menu items and group the message updates together with the current messaging prompts.

bdunne commented on lib/manageiq-appliance_console.rb:
```diff
@@ -39,6 +39,9 @@ def self.logger=(logger)
 require 'manageiq/appliance_console/external_auth_options'
 require 'manageiq/appliance_console/external_database_configuration'
 require 'manageiq/appliance_console/external_httpd_authentication'
+require 'manageiq/appliance_console/message_configuration'
```
Sort alphabetically?

bdunne replied:
Can this be merged into https://github.com/ManageIQ/manageiq-appliance_console/pull/130/files#diff-7a5a900c5a6ca9e861d311795135eaa7e7556074b5dc71399e90e12a7910b946R49 ?

jvlcek replied:
@bdunne Sorry that link is not clear. Where are you suggesting I merge this require into?

bdunne replied:
https://github.com/ManageIQ/manageiq-appliance_console/blob/481d9b37c73a44471b13ce5c5d7a8fce00b7429d/lib/manageiq-appliance_console.rb#L49

jvlcek replied:
@bdunne Alphabetize requires 👍 As discussed combining messaging_configuration with message_configure_client will be done in the future.

bdunne commented on lib/manageiq/appliance_console/message_configuration.rb:
```diff
@@ -0,0 +1,134 @@
+        @config_dir_path                   = Pathname.new(CONFIG_DIR)
+        @sample_config_dir_path            = Pathname.new(SAMPLE_CONFIG_DIR)
+
+        @tools_log4_properties_path        = @config_dir_path.join("", "tools-log4j.properties")
```
Why the empty quotes on these?

jvlcek replied:
@bdunne Good catch! Opps! Fixed by commit: https://github.com/ManageIQ/manageiq-appliance_console/pull/130/commits/dcc87dd8cf25254869dcdd0724dd44e491c7691c

bdunne commented on lib/manageiq/appliance_console/message_configuration.rb:
```diff
@@ -0,0 +1,134 @@
+      def create_tools_log_config
+        say(__method__.to_s.tr("_", " ").titleize)
+
+        FileUtils.cp(@tools_log4_properties_sample_path, @tools_log4_properties_path) unless file_found?(@tools_log4_properties_path)
```
You shouldn't need the `@` here since it's an `attr_reader` (same throughout the PR).

bdunne replied:
What do you think of `tools_log4_properties_path.exist?` (Pathname#exist?) rather than a new `file_found?` method?

Fryguy replied:
file_found? also logs a message (not saying it should, just that's why there's a separate method)

jvlcek replied:
Correct, @Fryguy the _file_found?_ method also logs the message letting  the user know the file already exists.
@bdunne I will remove the unnecessary `@`. I just like to leave them so when reading the code the scope of the variable is immediately  evident. But if it is more acceptable style I',m  OK with removing them.

Updated by commit: https://github.com/ManageIQ/manageiq-appliance_console/pull/130/commits/a1a2b4a5cf7ab94cfd249c654ae45ce324bdc34f

bdunne commented on lib/manageiq/appliance_console/message_configuration.rb:
```diff
@@ -0,0 +1,134 @@
+      def configure_messaging_yaml
+        say(__method__.to_s.tr("_", " ").titleize)
+
+        return true if file_found?(@messaging_yaml_path)
```
Do we need to return `true` here?  It doesn't look like it's used anywhere.  Also, if we don't return `true` here, the method will return an integer from `File.write`

jvlcek replied:
@bdunne Good point. Fixed by commit: https://github.com/ManageIQ/manageiq-appliance_console/pull/130/commits/a115b2753a13e4ae3bd1b8fe0769aa12bcc96fcf

bdunne commented on lib/manageiq/appliance_console/message_configuration.rb:
```diff
@@ -0,0 +1,134 @@
+        messaging_yaml["production"]["username"] = @username
+        messaging_yaml["production"]["password"] = @password
+
+        File.write(@messaging_yaml_path, messaging_yaml.to_yaml) unless file_found?(@messaging_yaml_path)
```
The `unless` shouldn't be needed since we checked a few lines above.

jvlcek replied:
@bdunne Good catch! Fixed by commit: https://github.com/ManageIQ/manageiq-appliance_console/pull/130/commits/5dcf698752f575400a690b9056d135ea22e7ec3b

bdunne commented on lib/manageiq/appliance_console/message_configuration.rb:
```diff
@@ -0,0 +1,134 @@
+      def remove_installed_files
+        say(__method__.to_s.tr("_", " ").titleize)
+
+        @installed_files.each { |f| FileUtils.rm_rf(f) if File.exist?(f) }
```
I don't think we need to check `if File.exist?` here

jvlcek replied:
Well it seems not. I knew FIleUtils.rm produces a traceback if the file does not exist. I didn't realize FileUtils.rm_rf() does not. @bdunne Once again **Good Catch!** 

Fixed by commit: https://github.com/ManageIQ/manageiq-appliance_console/pull/130/commits/327bca8ea1fab36dee03c3348a904d4af2ff200b

bdunne commented on lib/manageiq/appliance_console/message_configuration_server.rb:
```diff
@@ -0,0 +1,225 @@
+      def configure_firewall
+        say(__method__.to_s.tr("_", " ").titleize)
+
+        AwesomeSpawn.run!("firewall-cmd --add-port=9092/tcp --permanent") # insecure
```
Since we're going through the trouble of setting up SSL, I don't think we should listen on 9092 for insecure connections.

jvlcek replied:
@bdunne OK. I had meant to ask you opinion on this. 

Addressed by commit: https://github.com/ManageIQ/manageiq-appliance_console/pull/130/commits/834642c3ea9207676d43da07cf4eff73f94524fa

Fryguy commented on lib/manageiq/appliance_console/message_configuration.rb:
```diff
@@ -0,0 +1,134 @@
+
+        content.split("\n").each do |l|
+          l.gsub!("/", "\\/")
+          l.gsub!(/password=.*$/, "password=") # Remove the passord as it can have special characters that grep can not match.
```
Typo in comment - passord -> password

jvlcek replied:
@Fryguy Good catch! I ran spell checker. I wonder why it didn't flag this.

Fryguy commented on lib/manageiq/appliance_console/message_configuration_server.rb:
```diff
@@ -0,0 +1,225 @@
+        FileUtils.chmod(0o755, @keystore_dir_path)
+
+        # Generte a Java keystore and key pair, creating keystore.jks
+        AwesomeSpawn.run!("keytool -keystore #{@keystore_path} -alias localhost -validity 10000 -genkey -keyalg RSA -storepass #{@password} -keypass #{@password} -dname \"cn=#{@server_hostname}\" ")
```
AwesomeSpawn should be used to handle the individual parameters to ensure they are properly escaped.  For example, if someone puts a `|` character in their password, the current way is broken.  Prefer:

```suggestion
        AwesomeSpawn.run!("keytool", :params => {"-keystore" => @keystore_path, "-alias" => "localhost", "-validity" => 10000, "-genkey" => nil, "-keyalg" => "RSA", "-storepass" => @password, "-keypass" => @password, "-dname" => "cn=#{@server_hostname}")
```

jvlcek replied:
@Fryguy Thank you for you input. I've addressed this and the other AwesomeSpawn issue with commit: https://github.com/ManageIQ/manageiq-appliance_console/pull/130/commits/37575cb580cab9d0404b9640c9bad22706ae28e8

Fryguy commented on lib/manageiq/appliance_console/message_configuration_server.rb:
```diff
@@ -0,0 +1,225 @@
+        AwesomeSpawn.run!("keytool -keystore #{@keystore_path} -alias localhost -validity 10000 -genkey -keyalg RSA -storepass #{@password} -keypass #{@password} -dname \"cn=#{@server_hostname}\" ")
+
+        # Use openssl to create a new CA cert, creating ca-cert and ca-key
+        AwesomeSpawn.run!("PASSWORD='#{@password}' openssl req -new -x509 -keyout #{@ca_key_path} -out #{@ca_cert_path} -days 10000 -passout 'env:PASSWORD' -subj '/CN=something'")
```
And ENV can also be parameterized

```suggestion
        AwesomeSpawn.run!("openssl", :env => {"PASSWORD" => @password}, :params => ["req", "-new", "-x509", {"-keyout" => @ca_key_path, "-out" => @ca_cert_path, "-days" => 10000, "-passout" => "env:PASSWORD", "-subj" => "/CN=something"}])
```

"

jvlcek commented:
@bdunne and @Fryguy I believe I have addressed all the issues you have identified, so this PR should be ready to merge.

One issue @bdunne and I had discussed, which he agreed could be addressed in a follow on PR, will be tracked in [Issue 135](https://github.com/ManageIQ/manageiq-appliance_console/issues/135)

jvlcek commented on lib/manageiq/appliance_console/message_configuration.rb:
```diff
@@ -0,0 +1,134 @@
+        messaging_yaml["production"]["hostname"] = server_hostname
+        messaging_yaml["production"]["port"] = 9093
+        messaging_yaml["production"]["username"] = username
+        messaging_yaml["production"]["password"] = password
```
In looking at https://github.com/ManageIQ/manageiq-appliance_console/blob/master/lib/manageiq/appliance_console/messaging_configuration.rb#L85 the password should be encrypted here too.

jvlcek replied:
This is addressed by commit: https://github.com/ManageIQ/manageiq-appliance_console/pull/130/commits/86304c6da9c13f09867e41c9548ec25fb0108dd8

bdunne commented on lib/manageiq/appliance_console/message_configuration.rb:
```diff
@@ -0,0 +1,134 @@
+      LOGS_DIR                          = "#{BASE_DIR}/logs".freeze
+      CONFIG_DIR                        = "#{BASE_DIR}/config".freeze
+      SAMPLE_CONFIG_DIR                 = "#{BASE_DIR}/config-sample".freeze
+      MIQ_CONFIG_DIR                    = "/var/www/miq/vmdb/config".freeze
```
Should this use `ManageIQ::ApplianceConsole::RAILS_ROOT.join("config")`?

jvlcek replied:
Good point. I didn't even know that existed.
Addressed by commit: https://github.com/ManageIQ/manageiq-appliance_console/pull/130/commits/9bb79759c88dddb951dad05049193da90aaacbcc

bdunne commented on lib/manageiq/appliance_console/message_configuration.rb:
```diff
@@ -0,0 +1,134 @@
+      MIQ_CONFIG_DIR                    = "/var/www/miq/vmdb/config".freeze
+
+      def initialize(options = {})
+        @username, @password = nil
```
Is this needed if they're set on the next two lines?

jvlcek replied:
@bdunne Good catch.
Addressed by commit: https://github.com/ManageIQ/manageiq-appliance_console/pull/130/commits/2709f84a43f067589eb5cbca6c02e475d91a4307

bdunne commented on lib/manageiq/appliance_console/message_configuration.rb:
```diff
@@ -0,0 +1,134 @@
+      def create_tools_log_config
+        say(__method__.to_s.tr("_", " ").titleize)
+
+        FileUtils.cp(tools_log4_properties_sample_path, tools_log4_properties_path) unless file_found?(tools_log4_properties_path)
```
Did we miss this in the RPM spec?  If so, I'd prefer to fix it there.

jvlcek replied:
@bdunne Good idea.
Addressed by: https://github.com/ManageIQ/manageiq-rpm_build/pull/122
and commit: https://github.com/ManageIQ/manageiq-appliance_console/pull/130/commits/8545e96290fa0369a7693fb7f2f8cfb3824b49a5

bdunne commented on lib/manageiq/appliance_console/message_configuration.rb:
```diff
@@ -0,0 +1,134 @@
+            password=#{password} ;
+        CLIENT_PROPERTIES
+
+        File.write(client_properties_path, content) unless file_found?(client_properties_path)
```
Should we `return unless ...` earlier to avoid creating the heredoc or should we always write the file in case the values changed?

jvlcek replied:
It seems to be a balance of one extra line of code in the source or creating the heredoc.
I'll avoid the heredoc.
Addressed by commit: https://github.com/ManageIQ/manageiq-appliance_console/pull/130/commits/48209af0a4f77eff5d9286b3617d0dea94fad40d

bdunne commented on lib/manageiq/appliance_console/message_configuration_server.rb:
```diff
@@ -0,0 +1,224 @@
+                            messaging_yaml_path, LOGS_DIR, keystore_dir_path]
+      end
+
+      def ask_questions
```
Is there a difference between this method and https://github.com/ManageIQ/manageiq-appliance_console/pull/130/files#diff-b896847105d5f7861df9e75f19292936e53103942ce4a0187fa37cafd3671839R25 ? Should they be merged into the base class?

jvlcek replied:
Good catch! Thanks!

Addressed by commit: https://github.com/ManageIQ/manageiq-appliance_console/pull/130/commits/13470ad9c2c671cea76655a630e7c0752dee8e4e

bdunne commented on lib/manageiq/appliance_console/message_configuration_client.rb:
```diff
@@ -0,0 +1,118 @@
+      def initialize(options = {})
+        super(options)
+
+        @server_hostname, @server_username, @server_password = nil
```
Is this needed since they're defined below?

jvlcek replied:
Good point. I also fixed my incorrect fix of unnecessary initialization to nil in commit: https://github.com/ManageIQ/manageiq-appliance_console/pull/130/commits/9406e5bfb02693d7a32e0f853f6ef9a3a889a21b

bdunne commented on lib/manageiq/appliance_console/message_configuration_client.rb:
```diff
@@ -0,0 +1,118 @@
+      end
+
+      def installation_valid?
+        return true if LinuxAdmin::Rpm.list_installed.key?("kafka")
```
I un-WIP'd https://github.com/ManageIQ/manageiq-rpm_build/pull/91

jvlcek replied:
Ah, OK. Good idea.
Addressed by commit: https://github.com/ManageIQ/manageiq-appliance_console/pull/130/commits/c3df35e082ad7102e1f62be15cb5b08a46be334f

bdunne commented on lib/manageiq/appliance_console/message_configuration_client.rb:
```diff
@@ -0,0 +1,118 @@
+
+      def already_configured?
+        installed_file_found = false
+        installed_files.each do |f|
```
Could be simplified with `installed_files.any? ...`

jvlcek replied:
@bdunne I'm not sure how because I want to `say("Installed file #{f} found.")` to list every file found. It seems `installed_files.any?` returns `true` the first time the block returns `true`.

bdunne replied:
Ok

bdunne commented on lib/manageiq/appliance_console/message_configuration.rb:
```diff
@@ -0,0 +1,134 @@
+                  :messaging_yaml_sample_path, :messaging_yaml_path
+
+      BASE_DIR                          = "/opt/kafka".freeze
+      LOGS_DIR                          = "#{BASE_DIR}/logs".freeze
```
Do you know what's put here?  If it's server logs, we should probably move that to `/var/log`, but if it's message data that they're calling "logs" it's fine to stay here.

jvlcek replied:
@bdunne 

It seems to be server logs. I'm OK with reconfiguring Kafka to write to `/var/log` but it seems kafka's logging configuration is a bit complex so I if we decide these logs really should be relocated I'd like to tackle that in a follow on PR.
Are you OK with that?

bdunne replied:
Ok, followup sounds good.

bdunne commented on lib/manageiq/appliance_console/message_configuration_client.rb:
```diff
@@ -0,0 +1,118 @@
+
+        return if file_found?(truststore_path)
+
+        FileUtils.mkdir_p(keystore_dir_path)
```
Should this directory be included in the RPM so that it has the correct permissions and ownership?

bdunne replied:
Or should the two files in here just live in the config directory?

jvlcek replied:
@bdunne Good idea to create this directory in the RPM.

Updated the RPM spec to do just that in commit: https://github.com/ManageIQ/manageiq-rpm_build/pull/122/commits/a8b119569777798e03e28a4a9be87d416bca7a46

Updated here in commit: https://github.com/ManageIQ/manageiq-appliance_console/pull/130/commits/4d74dd3ab8fa320f25d0fcf40cc039a5fc71cbe9

bdunne commented on lib/manageiq/appliance_console/message_configuration_server.rb:
```diff
@@ -0,0 +1,224 @@
+        FileUtils.chmod(0o755, keystore_dir_path)
+
+        # Generte a Java keystore and key pair, creating keystore.jks
+        AwesomeSpawn.run!("keytool", :params => {"-keystore" => keystore_path, "-alias" => "localhost", "-validity" => 10_000, "-genkey" => nil, "-keyalg" => "RSA", "-storepass" => password, "-keypass" => password, "-dname" => "cn=#{server_hostname}"})
```
Are we able to put the password in an ENV var to help protect it?

jvlcek replied:
@bdunne I spent a fair amount of time exploring and experimenting with this and it does not see possible to pass passwords to keytool in an ENV var.

jvlcek commented on lib/manageiq/appliance_console/message_configuration_server.rb:
```diff
@@ -0,0 +1,224 @@
+        false
+      end
+
+      def already_configured?
```
Good catch. I'm sorry but I fixed this with the same commit where I removed checking for the kafka RPM.

Addressed in commit https://github.com/ManageIQ/manageiq-appliance_console/pull/130/commits/c3df35e082ad7102e1f62be15cb5b08a46be334f

jvlcek commented on lib/manageiq/appliance_console/message_configuration_server.rb:
```diff
@@ -0,0 +1,224 @@
+      def create_server_properties
+        say(__method__.to_s.tr("_", " ").titleize)
+
+        content = <<~SERVER_PROPERTIES
```
@bdunne Yes there are roughly 20 name=value pairs already in the server.properties file. I think doing it this way could be safer than hardcoding what is delivered in the file because future versions of kafka could be delivered with different content. Then we would have to update the content here too.
I propose we leave this as is.

bdunne commented:
Followup in https://github.com/ManageIQ/manageiq-appliance_console/issues/135
